src/builder/QuarkBuilder.sol

```diff
@@ -93,42 +96,80 @@ contract QuarkBuilder {
-            if (payment.isToken) {
-                // wrap around paycall
-                // TODO: need to embed price feed addresses for known tokens before we can do paycall.
-                // ^^^ look up USDC price feeds for each supported chain?
-                // we only need USDC/USD and only on chains 1 (mainnet) and 8453 (base mainnet).
-            } else {
-                quarkOperations[actionIndex++] = Actions.bridgeUSDC(
-                    Actions.BridgeUSDC({
-                        chainAccountsList: chainAccountsList,
-                        assetSymbol: transferIntent.assetSymbol,
-                        amount: transferIntent.amount,
-                        // where it comes from
-                        originChainId: 8453, // FIXME: originChainId
-                        sender: address(0), // FIXME: sender
-                        // where it goes
-                        destinationChainId: transferIntent.chainId,
-                        recipient: transferIntent.recipient,
-                        blockTimestamp: transferIntent.blockTimestamp
-                    })
-                );
-                // TODO: also append a Actions.Action to the actions array.
-                // See: BridgeUSDC TODO for returning a Actions.Action.
+            // Note: Assumes that the asset uses the same # of decimals on each chain
+            uint256 balanceOnDstChain =
+                Accounts.getBalanceOnChain(transferIntent.assetSymbol, transferIntent.chainId, chainAccountsList);
+            uint256 amountLeftToBridge = transferIntent.amount - balanceOnDstChain;
+
+            uint256 bridgeActionCount = 0;
+            // TODO: bridge routing logic (which bridge to prioritize, how many bridges?)
+            // Iterate chainAccountList and find upto 2 chains that can provide enough fund
+            // Backend can provide optimal routes by adjust the order in chainAccountList.
+            for (uint256 i = 0; i < chainAccountsList.length; ++i) {
+                if (amountLeftToBridge == 0) {
+                    break;
+                }
+
+                Accounts.ChainAccounts memory srcChainAccounts = chainAccountsList[i];
+                if (srcChainAccounts.chainId == transferIntent.chainId) {
+                    continue;
+                }
+
+                if (
+                    !BridgeRoutes.canBridge(srcChainAccounts.chainId, transferIntent.chainId, transferIntent.assetSymbol)
+                ) {
+                    continue;
+                }
+
+                Accounts.AssetPositions memory srcAssetPositions =
+                    Accounts.findAssetPositions(transferIntent.assetSymbol, srcChainAccounts.assetPositionsList);
+                Accounts.AccountBalance[] memory srcAccountBalances = srcAssetPositions.accountBalances;
+                // TODO: Make logic smarter. Currently, this uses a greedy algorithm.
+                for (uint256 j = 0; j < srcAccountBalances.length; ++j) {
```

Comment:
We should really order the actions by bridge priority, right?

Reply:
Right, we can sort the source chains by bridge priority.